### PR TITLE
Capitals, Primary Culture & Religion

### PR DESCRIPTION
**Pull request by Zemurin**
- State Religion conversion
- Primary Culture conversion
- Capital conversion for both extant and extinct eu4 imports.

### CODE REVIEW

File: EU4toV3/Source/Mappers/CultureMapper/CultureMapper.cpp [hunk 1]
```diff
@@ -259,6 +259,32 @@ void mappers::CultureMapper::registerKeys()
...
+	 const std::string& eu4religion,
+	 const std::string& v3state,
+	 const std::string& v3ownerTag)
+{
+	// This is a check where we ping for a v3 culture without having a clue what we're asking for exactly.
+	// 1. If the state is colonial, and we have a record of that eu4 culture having a neoculture in that colony,
+	//	then we return that neoculture.
+	// 2. If the state is colonial, and we don't have a record of eu4 culture having a neoculture in that colony,
+	//	it's possible we're a native country, and return regular match.
+	// 3. If the state is not colonial, return regular match same as 2.
```
Comment (Idhrendur) [inline EU4toV3/Source/Mappers/CultureMapper/CultureMapper.cpp:275]: Comments like this are better in the header.

**Review by Idhrendur — APPROVED** [review]

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 2]
```diff
@@ -78,26 +90,138 @@ void V3::Country::convertFromEU4Country(const ClayManager& clayManager)
...
+}
+
+void V3::Country::convertCulture(const ClayManager& clayManager,
+	 mappers::CultureMapper& cultureMapper,
+	 const EU4::CultureLoader& cultureLoader,
+	 const EU4::ReligionLoader& religionLoader)
+{
+	// Cultures can be tricky. We may have neocultures generated for our pops, meaning we need to account for being a neoculture country.
+	// We also can easily have a primary culture with 0 actual pops in our country.
+	// Again, as for Vic2, we're *ignoring* accepted cultures completely.
```
Comment (gawquon) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:138]: We absolutely are ignoring them, but worth thinking about later if a bunch of accepted cultures would increase the progressiveness of your Vic3 racial tolerance law. Probably just having the idea group that boosts number of accepted culture, whichever one that was.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 3]
```diff
@@ -78,26 +90,138 @@ void V3::Country::convertFromEU4Country(const ClayManager& clayManager)
...
+	{
+		processedData.capitalStateName = *historicalMatch;
+		return;
+	}
+
+	// still nothing?
+	if (processedData.capitalStateName.empty() && !substates.empty())
+		processedData.capitalStateName = substates[0]->getHomeStateName();
+
+	// TODO: Try anything harder. At least try to determine the majority of land?
```
Comment (gawquon) [inline EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:221]: SubState with most resource potential? That would save your capital from being Ob, Siberia when you have steppe territory.

File: EU4toV3/Source/V3World/ClayManager/State/Chunk.cpp [hunk 4]
```diff
@@ -41,6 +42,15 @@ std::map<std::string, double> V3::Chunk::calcOwnerWeights() const
 	return ownerWeights;
 }
 
+std::set<std::string> V3::Chunk::getKnownCapitals() const
+{
+	std::set<std::string> capitals;
+	for (const auto& wspd: weightedSourceProvinceData | std::views::keys)
+		if (!wspd.eu4Capitals.empty())
```
Comment (gawquon) [inline EU4toV3/Source/V3World/ClayManager/State/Chunk.cpp:49]: This is just a C++ question I have. If wspd.eu4Capitals is empty wouldn't .begin() == .end()? Making this equivalent with and without the if statement? Insert from 0 inclusive to 0 exclusive -> do nothing?

**Review by gawquon — COMMENTED** [review]
I'd approve but should that submodule fronter commit be there? Isn't that update handled on the main branch?

File: EU4toV3/Source/V3World/ClayManager/ClayManager.cpp [hunk 5]
```diff
@@ -260,6 +266,8 @@ std::vector<std::shared_ptr<V3::SubState>> V3::ClayManager::chunkToSubStatesTran
 	auto generatedSubStates = buildSubStates(sortedProvinces);
 
 	// update substates with metadata
+	std::shared_ptr<SubState> largestSubState;
+	double largestSubStateWeight = 0;
 	for (const auto& subState: generatedSubStates)
 	{
 		// for now, this is the simplest we can do. substateFactor is literally the amount of provinces / total provinces
```
[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV3/Source/V3World/ClayManager/ClayManager.cpp:273]: Apparently we don't do that yet? loadState() does that but it's never called. Landshare calc should be called as soon as all the substates are generated in [51%] clayManager.shoveRemainingProvincesIntoSubStates();

Also it should be recalced after the substates are squashed.

[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV3/Source/V3World/ClayManager/ClayManager.cpp:273]: I'll make a note on that TODO to do just this.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp (hunk 2 above)
Comment (Zemurin) [reply EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:138]: This is a good point but for later, when setting up accepted pops - compare against some internal country cultural tolerance tracker (stemming from eu4 ideas/reforms etc)  and allow some extra cultures if tolerant enough.

File: EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp (hunk 3 above)
Comment (Zemurin) [reply EU4toV3/Source/V3World/PoliticalManager/Country/Country.cpp:221]: Made a TODO inside to plug in resource calc at the same time as landshares.

File: EU4toV3/Source/V3World/ClayManager/State/Chunk.cpp (hunk 4 above)
Comment (Zemurin) [reply EU4toV3/Source/V3World/ClayManager/State/Chunk.cpp]: Thanx, fixed.

**Pull request merged**